src/OpenLoco/Paint/PaintVehicle.cpp

```diff
@@ -0,0 +1,198 @@
+#include "PaintVehicle.h"
+#include "../CompanyManager.h"
+#include "../Config.h"
+#include "../Graphics/Colour.h"
+#include "../Objects/ObjectManager.h"
+#include "../Objects/VehicleObject.h"
+#include "../Things/Vehicle.h"
+#include "Paint.h"
+
+using namespace OpenLoco::Interop;
+using namespace OpenLoco::Things::Vehicle;
+
+namespace OpenLoco::Paint
+{
+    // 0x00500160
+    const uint8_t _reversePitch[16]{
+        0, 5, 6, 7, 8, 1, 2, 3, 4, 10, 9, 12, 11, 0, 0, 0
+    };
+
+    // 0x004FFAE8
+    const uint32_t _ghostInvisImageFlags[2]{
+        Gfx::recolour(0, PaletteIndex::index_2C),
+        Gfx::recolourTranslucent(0, PaletteIndex::index_31),
+    };
+
+    // 0x004B0CFC
+    static void paintBogie(PaintSession& session, vehicle_bogie* bogie)
+    {
+        auto* vehObject = ObjectManager::get<vehicle_object>(bogie->object_id);
+        if (bogie->object_sprite_type == 0xFF)
+        {
+            return;
+        }
+
+        auto& sprite = vehObject->bogie_sprites[bogie->object_sprite_type];
+        uint8_t yaw = (bogie->sprite_yaw + (session.getRotation() << 4)) & 0x3F;
+        uint8_t pitch = bogie->sprite_pitch;
+
+        if (bogie->getFlags38() & Flags38::isReversed)
+        {
+            yaw ^= (1 << 5);
+            pitch = _reversePitch[bogie->sprite_pitch];
+        }
+        auto yawIndex = (yaw >> 1) & 0x1F;
+
+        switch (pitch)
+        {
+            case 0:
+            case 2:
+            case 9:
+            case 4:
+            case 11:
+                break;
+            default:
+                if (sprite.flags & BogieSpriteFlags::rotationalSymmetry)
+                {
+                    yawIndex ^= (1 << 4);
+                }
+                else
+                {
+                    yawIndex ^= (1 << 5);
+                }
+                break;
+        }
+        switch (pitch)
+        {
+            case 0:
+            {
+                if (sprite.flags & BogieSpriteFlags::rotationalSymmetry)
+                {
+                    yawIndex &= 0xF;
+                }
+                auto imageId = sprite.numRollSprites * yawIndex + bogie->var_46 + sprite.flatImageIds;
+                if (bogie->getFlags38() & Flags38::isGhost)
+                {
+                    imageId |= _ghostInvisImageFlags[Config::get().construction_marker];
```

Comment:
missing setting the session type to null

Reply:
Not a fan of changing the output of the Gfx::recolour() function afterwards. Would rather see a method that takes the imageId and *then* applies the recolour function (so that we can later increase the image limit)

Reply:
I've turned it into a function see latest version.